src/accelerate/utils/bnb.py

```diff
@@ -0,0 +1,416 @@
+            "You have a version of `bitsandbytes` that is not compatible with 4bit quantization,"
+            "make sure you have the latest version of `bitsandbytes` installed."
+        )
+
+    modules_on_cpu = []
+    if isinstance(device_map, dict) and len(device_map.keys()) > 1:
+        modules_on_cpu = [key for key, value in device_map.items() if value in ["disk", "cpu"]]
+        if len(modules_on_cpu) > 0 and not bnb_quantization_config.enable_fp32_cpu_offload:
+            raise ValueError(
+                "If you want to offload some keys to `cpu` or `disk`, you need to set "
+                " `enable_fp32_cpu_offload=True`. Note that these modules will not be "
+                " converted to 8-bit but kept in 32-bit."
+            )
+
+    # We keep some modules such as the lm_head in their original dtype for numerical stability reasons
+    if bnb_quantization_config.skip_modules is None:
+        bnb_quantization_config.skip_modules = get_keys_to_not_convert(model)
+
+    # add cpu modules to skip modules (after looking into the code on transformers, we don't really keep the cpu module in fp32)
+    bnb_quantization_config.skip_modules.extend(modules_on_cpu)
+    modules_to_not_convert = bnb_quantization_config.skip_modules
+    # We add the modules we want to keep in full precision
+
+    if bnb_quantization_config.keep_in_fp32_modules is None:
+        bnb_quantization_config.keep_in_fp32_modules = []
+    keep_in_fp32_modules = bnb_quantization_config.keep_in_fp32_modules
+    modules_to_not_convert.extend(keep_in_fp32_modules)
+
+    if model.device.type != "meta":
+        # quantization of an already loaded model
+        logger.warning(
+            "It is not recommended to quantize a loaded model. "
+            "The model should be instantiated under the `init_empty_weights` context manager."
+        )
+        model = replace_with_bnb_layers(model, bnb_quantization_config, modules_to_not_convert=modules_to_not_convert)
+        # convert param to the right dtype
+        dtype = bnb_quantization_config.torch_dtype
+        for name, param in model.state_dict().items():
+            if any(module_to_keep_in_fp32 in name for module_to_keep_in_fp32 in keep_in_fp32_modules):
+                param.to(torch.float32)
+                if param.dtype != torch.float32:
+                    name = name.replace(".weight", "").replace(".bias", "")
+                    param = getattr(model, name, None)
+                    if param is not None:
+                        param.to(torch.float32)
+            elif torch.is_floating_point(param):
+                param.to(dtype)
+        if model.device.type == "cuda":
+            # move everything to cpu in the first place because we can't do quantization if the weights are already on cuda
+            model.cuda(torch.cuda.current_device())
+            torch.cuda.empty_cache()
+        elif torch.cuda.is_available():
+            model.to(torch.cuda.current_device())
```

Comment:
I just want to understand this block, so you move back the model into a single GPU device - I was wondering what happens if the model is loaded across multiple GPUs ? Also the comment says move to cpu but it is calling `model.cuda()` (just want to know more about what that block is doing)

Reply:
I think that I have a better understanding now, this enables quantizing a model that is loaded independently from accelerate (e.g. without device_map=xxx), so `model.cuda` should recursively call that method to each submodule and quantize the weights, is that correct?

Reply:
Yes, this is how it works!